src/v/cloud_storage/remote_segment.cc

```diff
@@ -594,14 +594,18 @@ size_t remote_segment_batch_reader::produce(model::record_batch batch) {
 }
 
 ss::future<> remote_segment_batch_reader::stop() {
-    vlog(_ctxlog.debug, "remote_segment_batch_reader::close");
+    vlog(_ctxlog.debug, "remote_segment_batch_reader::stop");
     co_await _gate.close();
     if (_parser) {
-        vlog(
-          _ctxlog.debug, "remote_segment_batch_reader::close - parser-close");
+        vlog(_ctxlog.debug, "remote_segment_batch_reader::stop - parser-close");
         co_await _parser->close();
         _parser.reset();
     }
+    _stopped = true;
+}
+
+remote_segment_batch_reader::~remote_segment_batch_reader() noexcept {
+    vassert(_stopped, "Destroyed without stopping");
```

Comment:
nit: we can check if `_parser` is initialised instead of using a dedicated flag, the `_parser` field might not be initialised though

Reply:
Having a separate _stopped avoids a race: the class can be in a state where _parser is uninitialized, but another fiber is just entering read_some and will try and initialize it.  We need the full stop() to have been called (including waiting on the gate).